lib/src/command/upgrade.dart

```diff
@@ -44,14 +61,163 @@ class UpgradeCommand extends PubCommand {
       log.warning(log.yellow(
           'The --packages-dir flag is no longer used and does nothing.'));
     }
-    await entrypoint.acquireDependencies(SolveType.UPGRADE,
-        useLatest: argResults.rest,
-        dryRun: argResults['dry-run'],
-        precompile: argResults['precompile']);
+
+    if (argResults.wasParsed('breaking')) {
+      final useLatest = argResults.rest;
+
+      final rootPubspec = entrypoint.root.pubspec;
+      final resolvablePubspec =
+          _stripVersionConstraints(rootPubspec, useLatest: useLatest);
+
+      Map<String, PackageId> resolvablePackages;
+      await log.spinner('Resolving', () async {
+        resolvablePackages = await _tryResolve(resolvablePubspec);
+      }, condition: _shouldShowSpinner);
+
+      final dependencyChanges = <PackageRange, PackageId>{};
+      final allDependencies = [
+        ...rootPubspec.dependencies.values,
+        ...rootPubspec.devDependencies.values
+      ];
+
+      for (var package in allDependencies) {
+        final resolvedPackage = resolvablePackages[package.name];
+
+        if (resolvedPackage != null &&
+            !package.constraint.allows(resolvedPackage.version) &&
+            package.source is HostedSource) {
+          dependencyChanges[package] = resolvedPackage;
```

Comment:
Only upgrade the ones specified on the command line... ie. if I do `pub upgrade --breaking retry` this should not change the pubspec.yaml entry for `shelf`.

Reply:
I suppose we can contemplate making a different feature which says something like: `pub upgrade --raise-lower-bound` which does:
 (A) `pub upgrade` (as normally)
 (B) reads `pubspec.lock` and ensures for any package installed, the current version is the lower bound in `pubspec.yaml`.

I think this is a different feature, I'm also not sure we want to promote such feature :)

But I can see it useful in cases where you don't want to test against old versions of your dependencies, so you just update your package to have the latest versions of your dependencies as a requirement. Downside is that this make your package hard to use along side other packages, because the version constraints will be more tight.

Probably wouldn't be interesting as long as Flutter is pinning all of its dependencies.

Reply:
> Only upgrade the ones specified on the command line... ie. if I do pub upgrade --breaking retry this should not change the pubspec.yaml entry for shelf.

It should do that! It's a little subtle because when we remove the version upper bounds, we only do them for the ones specified in the command line, so when we resolve, there will never be a case where a package that was not specified in the command line will end up in a situation where `!package.constraint.allows(resolvedPackage.version)` is `true`.

Added a comment for that!